test/core/derivatives/implementation/base.py

```diff
@@ -1,23 +1,163 @@
-class DerivativesImplementation:
+"""Contains DerivativesImplementation, the base class for autograd and backpack."""
+from abc import ABC, abstractmethod
+
+from torch import Tensor
+
+
+class DerivativesImplementation(ABC):
     """Base class for autograd and BackPACK implementations."""
 
     def __init__(self, problem):
+        """Initialization.
+
+        Args:
+            problem: test problem
+        """
         self.problem = problem
 
-    def jac_mat_prod(self, mat):
+    @abstractmethod
+    def jac_mat_prod(self, mat: Tensor) -> Tensor:
+        """Product of jacobian and matrix.
```

Comment:
Vectorized application of output-input Jacobian.

Reply:
I think "vectorized" is not correct for AutogradDerivatives.
I would change it to
Product of output-input-Jacobian and a matrix.